File: backend/src/controllers/documentController.js
import asyncHandler from 'express-async-handler';
import Document from '../models/Document.js';
import Folder from '../models/Folder.js';
import Team from '../models/Team.js';
import { createNotifications, emitTeamUpdate } from '../utils/notificationService.js';
import { upload, cloudinary, CLOUDINARY_ENABLED } from '../config/cloudinary.js';
import path from 'path';
import fs from 'fs';

// ---------- helpers ----------

/**
 * Extract Cloudinary public_id from a full CDN URL.
 * e.g. https://res.cloudinary.com/{cloud}/raw/upload/v123/aurora-docs/1234-resume
 *   → "aurora-docs/1234-resume"
 */
const extractCloudinaryPublicId = (url) => {
    try {
        const match = url.match(/\/upload\/(?:v\d+\/)?(.+)$/);
        if (match) return match[1].replace(/\.[^/.]+$/, '');
    } catch (_) {}
    return null;
};

/**
 * Map a file extension to the Cloudinary resource_type needed for URL delivery.
 * PDFs and office docs must use 'raw' — accessing them via 'image' returns 401.
 */
const IMAGE_EXTS = ['jpg','jpeg','png','gif','webp','svg','bmp','ico','tiff','avif'];
const VIDEO_EXTS = ['mp4','avi','mov','webm','mkv','flv','wmv','m4v'];
const getCloudinaryResourceType = (filename) => {
    const ext = (filename || '').split('.').pop().toLowerCase();
    if (IMAGE_EXTS.includes(ext)) return 'image';
    if (VIDEO_EXTS.includes(ext)) return 'video';
    return 'raw';  // PDF, DOCX, XLSX, ZIP, etc.
};

/**
 * Delete a file from whichever storage it lives in.
 *   doc.cloudinaryId  → use stored public_id (preferred)
 *   doc.url = http(s) → extract public_id from URL (fallback for older records)
 *   doc.url = /uploads/... → delete from local disk
 */
const deleteStoredFile = async (doc) => {
    const isCloudinaryUrl = doc.url && doc.url.startsWith('http');

    if (isCloudinaryUrl && CLOUDINARY_ENABLED) {
        const publicId     = doc.cloudinaryId || extractCloudinaryPublicId(doc.url);
        const resourceType = getCloudinaryResourceType(doc.name || doc.url || '');
        if (publicId) {
            try {
                const result = await cloudinary.uploader.destroy(publicId, { resource_type: resourceType });
                console.log(`[Cloudinary] deleted public_id="${publicId}" type=${resourceType} result=${result.result}`);
            } catch (err) {
                console.error('[Cloudinary] delete error:', err.message);
            }
        } else {
            console.warn('[Cloudinary] could not determine public_id for URL:', doc.url);
        }
    } else if (!isCloudinaryUrl && doc.url && doc.url.startsWith('/uploads/')) {
        try {
            const filePath = path.join(path.resolve(), doc.url.replace(/^\//, ''));
            if (fs.existsSync(filePath)) {
                fs.unlinkSync(filePath);
                console.log(`[Local] deleted: ${filePath}`);
            }
        } catch (err) {
            console.error('[Local] delete error:', err.message);
        }
    }
};

// ---------- Controllers ----------

// @desc    Get documents and folders for a specific team
// @route   GET /api/documents?teamId=...&folderId=...
// @access  Private
const getDocuments = asyncHandler(async (req, res) => {
    const { teamId, folderId } = req.query;

    if (!teamId) {
        res.status(400);
        throw new Error('Team ID is required');
    }

    const team = await Team.findById(teamId);
    if (!team) {
        res.status(404);
        throw new Error('Team not found');
    }

    const userId = req.user.id.toString();
    const isMasterAdmin = req.user.role === 'master_admin';
    const isMember = isMasterAdmin
        || team.members.some(m => m && m.toString() === userId)
        || team.owner.toString() === userId;

    if (!isMember) {
        res.status(403);
        throw new Error('Not authorized to access this team files');
    }

    const currentFolderId = folderId && folderId !== 'null' ? folderId : null;

    const [folders, documents] = await Promise.all([
        Folder.find({ team: teamId, parentFolder: currentFolderId }).populate('createdBy', 'name'),
        Document.find({ team: teamId, folder: currentFolderId }).populate('uploadedBy', 'name'),
    ]);

    let currentFolder = null;
    if (currentFolderId) currentFolder = await Folder.findById(currentFolderId);

    res.json({ folders, documents, currentFolder });
});

// @desc    Upload file (Cloudinary if configured, local disk otherwise)
// @route   POST /api/documents/upload
// @access  Private
const uploadDocument = asyncHandler(async (req, res) => {
    await new Promise((resolve, reject) => {
        upload(req, res, (err) => {
            if (err) return reject(err);
            resolve();
        });
    }).catch((err) => {
        res.status(400);
        throw new Error(err.message || 'File upload failed');
    });

    const { teamId, folderId } = req.body;

    if (!req.file) {
        res.status(400);
        throw new Error('No file uploaded');
    }

    const team = await Team.findById(teamId);
    if (!team) {
        await deleteStoredFile({ cloudinaryId: req.file.filename, url: req.file.path });
        res.status(404);
        throw new Error('Team not found');
    }

    const userId = req.user.id.toString();
    const isMasterAdmin = req.user.role === 'master_admin';
    const isMember = isMasterAdmin
        || team.members.some(m => m && m.toString() === userId)
        || team.owner.toString() === userId;

    if (!isMember) {
        await deleteStoredFile({ cloudinaryId: req.file.filename, url: req.file.path });
        res.status(403);
        throw new Error('Not authorized to upload to this team');
    }

    // Cloudinary mode → req.file.path = https CDN URL, req.file.filename = public_id
    // Local mode      → req.file.path = absolute disk path, build relative URL
    let storedUrl;
    let cloudinaryId = null;

    if (CLOUDINARY_ENABLED && req.file.path && req.file.path.startsWith('http')) {
        storedUrl    = req.file.path;
        cloudinaryId = req.file.filename;
    } else {
        storedUrl = `/uploads/${req.file.filename}`;
    }

    const document = await Document.create({
        team: teamId,
        folder: folderId && folderId !== 'null' ? folderId : null,
        uploadedBy: req.user.id,
        user: req.user.id,
        name: req.file.originalname,
        type: req.file.originalname.split('.').pop().toLowerCase(),
        size: req.file.size
            ? `${(req.file.size / (1024 * 1024)).toFixed(2)} MB`
            : 'Unknown',
        url: storedUrl,
        cloudinaryId,
    });

    const io = req.app.get('socketio');
    const recipientIds = [...team.members, team.owner]
        .filter(id => id != null)
        .map(id => id.toString())
        .filter(id => id !== req.user.id.toString());

    if (recipientIds.length > 0) {
        await createNotifications(recipientIds, {
            title: 'New Document Uploaded',
            description: `"${document.name}" has been uploaded to the team library`,
            type: 'document_shared',
            sender: req.user.id,
            link: '/document-share',
        }, io);
    }

    emitTeamUpdate(io, teamId, 'DOCUMENT_UPLOAD');

    const populatedDoc = await Document.findById(document._id).populate('uploadedBy', 'name');
    res.status(201).json(populatedDoc);
});

// @desc    Create Folder
// @route   POST /api/documents/folder
// @access  Private
const createFolder = asyncHandler(async (req, res) => {
    const { name, teamId, parentFolderId } = req.body;
    if (!name || !teamId) {
        res.status(400);
        throw new Error('Name and Team ID required');
    }

    const folder = await Folder.create({
        name,
        team: teamId,
        parentFolder: parentFolderId || null,
        createdBy: req.user.id,
    });

    const io = req.app.get('socketio');
    emitTeamUpdate(io, teamId, 'FOLDER_CREATE');
    res.status(201).json(folder);
});

// @desc    Rename Folder (Head/Admin only)
// @route   PUT /api/documents/folder/:id
// @access  Private
const renameFolder = asyncHandler(async (req, res) => {
    const { name } = req.body;
    const folder = await Folder.findById(req.params.id);
    if (!folder) {
        res.status(404);
        throw new Error('Folder not found');
    }

    const isMasterAdmin = req.user.role === 'master_admin';
    const isHead = isMasterAdmin || req.user.role === 'team_head' || req.user.role === 'admin';
    if (!isHead) {
        res.status(403);
        throw new Error('Only Team Heads can rename folders');
    }

    folder.name = name || folder.name;
    await folder.save();

    const io = req.app.get('socketio');
    emitTeamUpdate(io, folder.team, 'FOLDER_RENAME');
    res.json(folder);
});

// @desc    Delete Folder (Head/Admin only) — recursive
// @route   DELETE /api/documents/folder/:id
// @access  Private
const deleteFolder = asyncHandler(async (req, res) => {
    const folder = await Folder.findById(req.params.id);
    if (!folder) {
        res.status(404);
        throw new Error('Folder not found');
    }

    const isMasterAdmin = req.user.role === 'master_admin';
    const isHead = isMasterAdmin || req.user.role === 'team_head' || req.user.role === 'admin';
    if (!isHead) {
        res.status(403);
        throw new Error('Only Team Heads can delete folders');
    }

    const teamId = folder.team;

    const deleteFolderRecursive = async (folderId) => {
        const subfolders = await Folder.find({ parentFolder: folderId });
        for (const sub of subfolders) await deleteFolderRecursive(sub._id);

        const docs = await Document.find({ folder: folderId });
        for (const doc of docs) {
            await deleteStoredFile(doc);
            await doc.deleteOne();
        }
        await Folder.findByIdAndDelete(folderId);
    };

    const io = req.app.get('socketio');
    const team = await Team.findById(folder.team);
    if (team) {
        const recipientIds = [...team.members, team.owner]
            .filter(id => id != null)
            .map(id => id.toString())
            .filter(id => id !== req.user.id.toString());

        if (recipientIds.length > 0) {
            await createNotifications(recipientIds, {
                title: 'Folder Removed',
                description: `${req.user.name} deleted the folder "${folder.name}" and its contents`,
                type: 'document_shared',
                sender: req.user.id,
                link: '/document-share',
            }, io);
        }
    }

    await deleteFolderRecursive(folder._id);
    emitTeamUpdate(io, teamId, 'FOLDER_DELETE');
    res.json({ message: 'Folder deleted' });
});

// @desc    Download document (auth-gated)
// @route   GET /api/documents/download/:id
// @access  Private
//
// For Cloudinary files: verify auth, return JSON { downloadUrl } with fl_attachment.
//   The browser fetches directly from CDN — no server-side proxy needed.
// For local files: stream from disk with Content-Disposition: attachment.
const downloadDocument = asyncHandler(async (req, res) => {
    const document = await Document.findById(req.params.id);
    if (!document) {
        res.status(404);
        throw new Error('Document not found');
    }

    const team = await Team.findById(document.team);
    if (!team) {
        res.status(404);
        throw new Error('Team not found');
    }

    const userId      = req.user.id.toString();
    const isMasterAdmin = req.user.role === 'master_admin';
    const isHead      = isMasterAdmin || req.user.role === 'team_head';
    const isMember    = isMasterAdmin
        || team.members.some(m => m && m.toString() === userId)
        || team.owner.toString() === userId;

    if (!isMember) {
        res.status(403);
        throw new Error('Not authorized to download this file');
    }

    if (!document.isDownloadable && !isHead) {
        res.status(403);
        throw new Error('Downloads have been restricted for this file by the team head');
    }

    const fileName  = document.name;
    const storedUrl = document.url || '';

    console.log(`[Download] doc=${document._id}  url=${storedUrl}`);

    // ── CASE 1: Absolute URL (Cloudinary CDN) ────────────────────────────────
    if (storedUrl.startsWith('http')) {
        const isLocalhost = /https?:\/\/(localhost|127\.0\.0\.1)(:\d+)?/.test(storedUrl);

        if (isLocalhost) {
            // Localhost URL stored by mistake — strip host, serve from disk
            const urlPath  = storedUrl.replace(/^https?:\/\/[^/]+/, '');
            const diskPath = path.join(path.resolve(), urlPath.replace(/^\//, ''));
            if (fs.existsSync(diskPath)) {
                const enc = encodeURIComponent(fileName);
                res.setHeader('Content-Disposition', `attachment; filename="${fileName}"; filename*=UTF-8''${enc}`);
                res.setHeader('Content-Type', 'application/octet-stream');
                res.setHeader('Access-Control-Expose-Headers', 'Content-Disposition');
                fs.createReadStream(diskPath).pipe(res);
                // Self-heal DB record
                Document.findByIdAndUpdate(document._id, { url: urlPath, cloudinaryId: null }).catch(() => {});
            } else {
                res.status(404).json({ message: 'File not found on disk. Please re-upload.' });
            }
            return;
        }

        // Real Cloudinary URL.
        // Build the download URL via the Cloudinary SDK so the resource_type is always
        // correct (image/video/raw). Accessing a raw PDF via /image/upload/ gives 401.
        const publicId = document.cloudinaryId || extractCloudinaryPublicId(storedUrl);
        const resourceType = getCloudinaryResourceType(fileName);

        let downloadUrl;
        if (publicId && CLOUDINARY_ENABLED) {
            // SDK-generated URL — guaranteed correct resource_type + fl_attachment
            downloadUrl = cloudinary.url(publicId, {
                resource_type: resourceType,
                secure: true,
                flags: 'attachment',
            });
            console.log(`[Download] SDK URL (${resourceType}): ${downloadUrl}`);
        } else {
            // Fallback: patch the stored URL resource type from the extension
            const correctType = resourceType; // 'image' | 'video' | 'raw'
            downloadUrl = storedUrl
                .replace(/\/(image|video|raw)\/upload\//, `/${correctType}/upload/`)
                .replace('/upload/', '/upload/fl_attachment/');
            console.log(`[Download] Patched URL: ${downloadUrl}`);
        }

        return res.json({ downloadUrl, fileName });
    }

    // ── CASE 2: Local disk (/uploads/...) ────────────────────────────────────
    const filePath = path.join(path.resolve(), storedUrl.replace(/^\//, ''));
    if (!fs.existsSync(filePath)) {
        return res.status(404).json({
            message: 'File not found on server. It may have been lost after a restart. Please re-upload.',
        });
    }

    const enc = encodeURIComponent(fileName);
    res.setHeader('Content-Disposition', `attachment; filename="${fileName}"; filename*=UTF-8''${enc}`);
    res.setHeader('Content-Type', 'application/octet-stream');
    res.setHeader('Access-Control-Expose-Headers', 'Content-Disposition');

    const fileStream = fs.createReadStream(filePath);
    fileStream.on('error', (e) => {
        console.error('[Download] stream error:', e.message);
        if (!res.headersSent) res.status(500).json({ message: 'Error reading file' });
    });
    fileStream.pipe(res);
});

// @desc    Delete document
// @route   DELETE /api/documents/:id
// @access  Private (uploader, head, master_admin)
const deleteDocument = asyncHandler(async (req, res) => {
    const document = await Document.findById(req.params.id);
    if (!document) {
        res.status(404);
        throw new Error('Document not found');
    }

    const isUploader  = document.uploadedBy.toString() === req.user.id.toString();
    const isMasterAdmin = req.user.role === 'master_admin';
    const isHead      = isMasterAdmin || req.user.role === 'team_head' || req.user.role === 'admin';

    if (!isUploader && !isHead) {
        res.status(403);
        throw new Error('Not authorized to delete this file');
    }

    const teamId = document.team;
    await deleteStoredFile(document);
    await document.deleteOne();

    const team = await Team.findById(teamId);
    const io   = req.app.get('socketio');
    if (team) {
        const recipientIds = [...team.members, team.owner]
            .filter(id => id != null)
            .map(id => id.toString())
            .filter(id => id !== req.user.id.toString());

        if (recipientIds.length > 0) {
            await createNotifications(recipientIds, {
                title: 'Document Removed',
                description: `${req.user.name} deleted "${document.name}"`,
                type: 'document_shared',
                sender: req.user.id,
                link: '/document-share',
            }, io);
        }
    }

    emitTeamUpdate(io, teamId, 'DOCUMENT_DELETE');
    res.json({ id: req.params.id });
});

// @desc    Toggle isDownloadable flag
// @route   PATCH /api/documents/:id/downloadable
// @access  Private (uploader, head, master_admin)
const toggleDownloadable = asyncHandler(async (req, res) => {
    const document = await Document.findById(req.params.id);
    if (!document) {
        res.status(404);
        throw new Error('Document not found');
    }

    const isMasterAdmin = req.user.role === 'master_admin';
    const isHead        = isMasterAdmin || req.user.role === 'team_head';
    const isUploader    = document.uploadedBy.toString() === req.user.id.toString();

    if (!isHead && !isUploader) {
        res.status(403);
        throw new Error('Only Team Heads, Master Admins, or the uploader can change download permissions');
    }

    document.isDownloadable = !document.isDownloadable;
    await document.save();

    const io = req.app.get('socketio');
    emitTeamUpdate(io, document.team, 'DOCUMENT_UPDATE');
    res.json({ _id: document._id, isDownloadable: document.isDownloadable });
});

export {
    getDocuments,
    uploadDocument,
    downloadDocument,
    deleteDocument,
    createFolder,
    renameFolder,
    deleteFolder,
    toggleDownloadable,
};

File: backend/controllers/document_controller.go
package controllers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"backend/config"
	"backend/middleware"
	"backend/models"
	"backend/utils"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// GetDocuments returns documents and folders for a team and optional folder
// GET /api/documents?teamId=...&folderId=...
func GetDocuments(w http.ResponseWriter, r *http.Request) {
	user := middleware.GetUserFromContext(r)
	if user == nil {
		utils.WriteError(w, http.StatusUnauthorized, "Not authorized")
		return
	}

	teamIDStr := r.URL.Query().Get("teamId")
	if teamIDStr == "" {
		utils.WriteError(w, http.StatusBadRequest, "Team ID is required")
		return
	}

	teamObjID, err := primitive.ObjectIDFromHex(teamIDStr)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "Invalid team ID format")
		return
	}

	folderIDStr := r.URL.Query().Get("folderId")
	var folderFilter any = nil
	if folderIDStr != "" && folderIDStr != "null" && folderIDStr != "root" {
		fID, err := primitive.ObjectIDFromHex(folderIDStr)
		if err == nil {
			folderFilter = fID
		}
	}

	docColl := config.DB.Collection("documents")
	folderColl := config.DB.Collection("folders")

	docFilter := bson.M{
		"team":   teamObjID,
		"folder": folderFilter,
	}

	docCursor, err := docColl.Find(r.Context(), docFilter, options.Find().SetSort(bson.M{"createdAt": -1}))
	if err != nil {
		utils.WriteError(w, http.StatusInternalServerError, "Failed to load documents")
		return
	}
	defer docCursor.Close(r.Context())

	var documents []models.Document
	_ = docCursor.All(r.Context(), &documents)

	// Fetch folders
	folderFilterQuery := bson.M{
		"team":         teamObjID,
		"parentFolder": folderFilter,
	}
	folderCursor, err := folderColl.Find(r.Context(), folderFilterQuery, options.Find().SetSort(bson.M{"name": 1}))
	if err != nil {
		utils.WriteError(w, http.StatusInternalServerError, "Failed to load folders")
		return
	}
	defer folderCursor.Close(r.Context())

	var folders []models.Folder
	_ = folderCursor.All(r.Context(), &folders)

	utils.WriteJSON(w, http.StatusOK, map[string]any{
		"documents": documents,
		"folders":   folders,
	})
}

// UploadDocument handles file upload (PDFs, images, docs) with full inline viewing support
// POST /api/documents/upload
func UploadDocument(w http.ResponseWriter, r *http.Request) {
	user := middleware.GetUserFromContext(r)
	if user == nil {
		utils.WriteError(w, http.StatusUnauthorized, "Not authorized")
		return
	}

	// 50 MB max
	if err := r.ParseMultipartForm(50 << 20); err != nil {
		utils.WriteError(w, http.StatusBadRequest, "File size too large or invalid form")
		return
	}

	file, header, err := r.FormFile("file")
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "No file uploaded")
		return
	}
	_ = file.Close()

	teamIDStr := r.FormValue("teamId")
	if teamIDStr == "" {
		utils.WriteError(w, http.StatusBadRequest, "Team ID is required")
		return
	}

	teamObjID, err := primitive.ObjectIDFromHex(teamIDStr)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "Invalid team ID format")
		return
	}

	folderIDStr := r.FormValue("folderId")
	var folderObjID *primitive.ObjectID
	if folderIDStr != "" && folderIDStr != "null" && folderIDStr != "root" {
		if fID, err := primitive.ObjectIDFromHex(folderIDStr); err == nil {
			folderObjID = &fID
		}
	}

	// Save to local storage engine
	fileURL, fileSize, err := config.SaveUploadedFile(header)
	if err != nil {
		utils.WriteError(w, http.StatusInternalServerError, fmt.Sprintf("Failed to save file: %v", err))
		return
	}

	sizeFormatted := formatFileSize(fileSize)
	ext := strings.TrimPrefix(filepath.Ext(header.Filename), ".")
	if ext == "" {
		ext = "file"
	}

	now := time.Now()
	newDoc := models.Document{
		ID:             primitive.NewObjectID(),
		User:           user.ID,
		UploadedBy:     user.ID,
		Name:           header.Filename,
		Type:           ext,
		Size:           sizeFormatted,
		URL:            fileURL,
		Folder:         folderObjID,
		Team:           teamObjID,
		IsDownloadable: true,
		CreatedAt:      now,
		UpdatedAt:      now,
	}

	docColl := config.DB.Collection("documents")
	_, err = docColl.InsertOne(r.Context(), newDoc)
	if err != nil {
		utils.WriteError(w, http.StatusInternalServerError, "Failed to save document record")
		return
	}

	utils.EmitTeamUpdate(teamObjID.Hex(), "DOCUMENT_UPLOAD", nil)

	utils.WriteJSON(w, http.StatusCreated, newDoc)
}

// ViewDocument serves a document (especially PDF) inline in the browser for instant preview
// GET /api/documents/:id/view
func ViewDocument(w http.ResponseWriter, r *http.Request) {
	idStr := r.PathValue("id")
	docObjID, err := primitive.ObjectIDFromHex(idStr)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "Invalid document ID")
		return
	}

	docColl := config.DB.Collection("documents")
	var doc models.Document
	if err := docColl.FindOne(r.Context(), bson.M{"_id": docObjID}).Decode(&doc); err != nil {
		utils.WriteError(w, http.StatusNotFound, "Document not found")
		return
	}

	// Check if local file
	if strings.HasPrefix(doc.URL, "/uploads/") {
		relPath := strings.TrimPrefix(doc.URL, "/uploads/")
		fullPath := filepath.Join(config.UploadsDir, filepath.Clean(relPath))
		config.ServeFileInline(w, r, fullPath, doc.Name)
		return
	}

	// If remote URL, redirect
	http.Redirect(w, r, doc.URL, http.StatusTemporaryRedirect)
}

// DownloadDocument serves a document as an attachment
// GET /api/documents/:id/download
func DownloadDocument(w http.ResponseWriter, r *http.Request) {
	idStr := r.PathValue("id")
	docObjID, err := primitive.ObjectIDFromHex(idStr)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "Invalid document ID")
		return
	}

	docColl := config.DB.Collection("documents")
	var doc models.Document
	if err := docColl.FindOne(r.Context(), bson.M{"_id": docObjID}).Decode(&doc); err != nil {
		utils.WriteError(w, http.StatusNotFound, "Document not found")
		return
	}

	if !doc.IsDownloadable {
		utils.WriteError(w, http.StatusForbidden, "Downloads restricted by administrator")
		return
	}

	if strings.HasPrefix(doc.URL, "/uploads/") {
		relPath := strings.TrimPrefix(doc.URL, "/uploads/")
		fullPath := filepath.Join(config.UploadsDir, filepath.Clean(relPath))
		w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, doc.Name))
		http.ServeFile(w, r, fullPath)
		return
	}

	http.Redirect(w, r, doc.URL, http.StatusTemporaryRedirect)
}

// CreateFolder creates a new folder in a team
// POST /api/documents/folder
func CreateFolder(w http.ResponseWriter, r *http.Request) {
	user := middleware.GetUserFromContext(r)
	if user == nil {
		utils.WriteError(w, http.StatusUnauthorized, "Not authorized")
		return
	}

	var req struct {
		Name           string `json:"name"`
		TeamID         string `json:"teamId"`
		ParentFolder   string `json:"parentFolder"`
		ParentFolderID string `json:"parentFolderId"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil || req.Name == "" || req.TeamID == "" {
		utils.WriteError(w, http.StatusBadRequest, "Folder name and team ID are required")
		return
	}

	teamObjID, err := primitive.ObjectIDFromHex(req.TeamID)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "Invalid team ID")
		return
	}

	parentStr := req.ParentFolder
	if parentStr == "" {
		parentStr = req.ParentFolderID
	}

	var parentObjID *primitive.ObjectID
	if parentStr != "" && parentStr != "null" && parentStr != "root" {
		if pID, err := primitive.ObjectIDFromHex(parentStr); err == nil {
			parentObjID = &pID
		}
	}

	now := time.Now()
	newFolder := models.Folder{
		ID:           primitive.NewObjectID(),
		Name:         req.Name,
		Team:         teamObjID,
		CreatedBy:    user.ID,
		ParentFolder: parentObjID,
		CreatedAt:    now,
		UpdatedAt:    now,
	}

	folderColl := config.DB.Collection("folders")
	_, err = folderColl.InsertOne(r.Context(), newFolder)
	if err != nil {
		utils.WriteError(w, http.StatusInternalServerError, "Failed to create folder")
		return
	}

	utils.EmitTeamUpdate(teamObjID.Hex(), "FOLDER_CREATE", nil)

	utils.WriteJSON(w, http.StatusCreated, newFolder)
}

// RenameFolder renames an existing folder
// PUT /api/documents/folder/:id
func RenameFolder(w http.ResponseWriter, r *http.Request) {
	user := middleware.GetUserFromContext(r)
	if user == nil {
		utils.WriteError(w, http.StatusUnauthorized, "Not authorized")
		return
	}

	idStr := r.PathValue("id")
	folderObjID, err := primitive.ObjectIDFromHex(idStr)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "Invalid folder ID")
		return
	}

	var req struct {
		Name string `json:"name"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil || req.Name == "" {
		utils.WriteError(w, http.StatusBadRequest, "New folder name is required")
		return
	}

	folderColl := config.DB.Collection("folders")
	var folder models.Folder
	if err := folderColl.FindOne(r.Context(), bson.M{"_id": folderObjID}).Decode(&folder); err != nil {
		utils.WriteError(w, http.StatusNotFound, "Folder not found")
		return
	}

	isHead := user.Role == models.RoleMasterAdmin || user.Role == models.RoleTeamHead || user.Role == models.RoleAdmin || string(user.Role) == "master"
	if !isHead {
		utils.WriteError(w, http.StatusForbidden, "Only Team Heads can rename folders")
		return
	}

	folder.Name = req.Name
	folder.UpdatedAt = time.Now()
	_, _ = folderColl.UpdateOne(r.Context(), bson.M{"_id": folderObjID}, bson.M{
		"$set": bson.M{"name": folder.Name, "updatedAt": folder.UpdatedAt},
	})

	utils.EmitTeamUpdate(folder.Team.Hex(), "FOLDER_RENAME", nil)
	utils.WriteJSON(w, http.StatusOK, folder)
}

// DeleteFolder recursively deletes a folder and its documents
// DELETE /api/documents/folder/:id
func DeleteFolder(w http.ResponseWriter, r *http.Request) {
	user := middleware.GetUserFromContext(r)
	if user == nil {
		utils.WriteError(w, http.StatusUnauthorized, "Not authorized")
		return
	}

	idStr := r.PathValue("id")
	folderObjID, err := primitive.ObjectIDFromHex(idStr)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "Invalid folder ID")
		return
	}

	folderColl := config.DB.Collection("folders")
	docColl := config.DB.Collection("documents")

	var folder models.Folder
	if err := folderColl.FindOne(r.Context(), bson.M{"_id": folderObjID}).Decode(&folder); err != nil {
		utils.WriteError(w, http.StatusNotFound, "Folder not found")
		return
	}

	isHead := user.Role == models.RoleMasterAdmin || user.Role == models.RoleTeamHead || user.Role == models.RoleAdmin || string(user.Role) == "master"
	if !isHead {
		utils.WriteError(w, http.StatusForbidden, "Only Team Heads can delete folders")
		return
	}

	// Recursive deletion helper
	var deleteRecursive func(fID primitive.ObjectID)
	deleteRecursive = func(fID primitive.ObjectID) {
		cursor, err := folderColl.Find(r.Context(), bson.M{"parentFolder": fID})
		if err == nil {
			var subs []models.Folder
			_ = cursor.All(r.Context(), &subs)
			cursor.Close(r.Context())
			for _, sub := range subs {
				deleteRecursive(sub.ID)
			}
		}

		docCursor, err := docColl.Find(r.Context(), bson.M{"folder": fID})
		if err == nil {
			var docs []models.Document
			_ = docCursor.All(r.Context(), &docs)
			docCursor.Close(r.Context())
			for _, d := range docs {
				_ = config.DeleteFile(d.URL)
				_, _ = docColl.DeleteOne(r.Context(), bson.M{"_id": d.ID})
			}
		}
		_, _ = folderColl.DeleteOne(r.Context(), bson.M{"_id": fID})
	}

	deleteRecursive(folderObjID)

	utils.EmitTeamUpdate(folder.Team.Hex(), "FOLDER_DELETE", nil)
	utils.WriteJSON(w, http.StatusOK, map[string]any{"id": idStr, "message": "Folder deleted"})
}


// DeleteDocument deletes a document
// DELETE /api/documents/:id
func DeleteDocument(w http.ResponseWriter, r *http.Request) {
	idStr := r.PathValue("id")
	docObjID, err := primitive.ObjectIDFromHex(idStr)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "Invalid document ID")
		return
	}

	docColl := config.DB.Collection("documents")
	var doc models.Document
	if err := docColl.FindOne(r.Context(), bson.M{"_id": docObjID}).Decode(&doc); err != nil {
		utils.WriteError(w, http.StatusNotFound, "Document not found")
		return
	}

	// Delete from storage
	_ = config.DeleteFile(doc.URL)

	_, _ = docColl.DeleteOne(r.Context(), bson.M{"_id": docObjID})
	utils.EmitTeamUpdate(doc.Team.Hex(), "DOCUMENT_DELETE", nil)

	utils.WriteJSON(w, http.StatusOK, map[string]any{"id": idStr})
}

// ToggleDownloadable toggles download permission
// PATCH /api/documents/:id/downloadable
func ToggleDownloadable(w http.ResponseWriter, r *http.Request) {
	idStr := r.PathValue("id")
	docObjID, err := primitive.ObjectIDFromHex(idStr)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "Invalid document ID")
		return
	}

	var req struct {
		IsDownloadable bool `json:"isDownloadable"`
	}
	_ = json.NewDecoder(r.Body).Decode(&req)

	docColl := config.DB.Collection("documents")
	_, err = docColl.UpdateOne(r.Context(), bson.M{"_id": docObjID}, bson.M{"$set": bson.M{"isDownloadable": req.IsDownloadable}})
	if err != nil {
		utils.WriteError(w, http.StatusInternalServerError, "Failed to update download permissions")
		return
	}

	utils.WriteJSON(w, http.StatusOK, map[string]any{"isDownloadable": req.IsDownloadable})
}

func formatFileSize(bytes int64) string {
	const unit = 1024
	if bytes < unit {
		return fmt.Sprintf("%d B", bytes)
	}
	div, exp := int64(unit), 0
	for n := bytes / unit; n >= unit; n /= unit {
		div *= unit
		exp++
	}
	return fmt.Sprintf("%.1f %cB", float64(bytes)/float64(div), "KMGTPE"[exp])
}

// Helper to parse int with fallback
func parseIntDefault(s string, def int) int {
	v, err := strconv.Atoi(s)
	if err != nil {
		return def
	}
	return v
}
